flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/TableFunction.scala

```diff
@@ -86,12 +90,79 @@ abstract class TableFunction[T] extends UserDefinedFunction {
     * @return [[Expression]] in form of a [[TableFunctionCall]]
     */
   final def apply(params: Expression*)(implicit typeInfo: TypeInformation[T]): Expression = {
-    val resultType = if (getResultType == null) {
-      typeInfo
+    buildTableFunctionCall(getClass.getSimpleName, typeInfo, params: _*)
+  }
+
+  /**
+    * Build a TableFunctionCall, a name and a sequence of params will determine a unique
+    * [[TableFunctionCall]]
+    *
+    * @param name function name
+    * @param implicitResultType If no result type returned, it will use this type.
+    * @param params The input expressions
+    * @return A unique [[TableFunctionCall]]
+    */
+  private[table] def buildTableFunctionCall(name: String,
```

Comment:
Can we move this to a util class? Since this is a user facing API class, we don't want to expose such kind of thing.